### PR TITLE
Implemented global access wideners

### PR DESCRIPTION
**Pull request by shartte**
Implemented global access wideners, and adapted a visitor-styled interface for more composability.
This also adds a new strict mode for the reader, which will reject class-names separated by periods rather than forward slashes, to catch a common mistake.

Example with new syntax:
```
accessWidener	v2	somenamespace
transitive-accessible	class	pkg/Class
transitive-extendable	class	pkg/Class
transitive-accessible	method	pkg/Class	method	()V
transitive-extendable	method	pkg/Class	method	()V
transitive-accessible	field	pkg/Class	field	I
transitive-mutable      field	pkg/Class	field	I
```

Code coverage is @ 94% lines.

### CODE REVIEW

File: src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java
Comment (modmuss50) [inline src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java:88]: `transitive` would be a better prefix.

File: src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java
Comment (modmuss50) [inline src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java:97]: I would move all of this out to be done later, concenrating on one thing at once is better imo.

File: src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java
Comment (modmuss50) [inline src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java:89]: This should be `transitive-<access>` over `transitive <access>

**Review by modmuss50 — CHANGES_REQUESTED** [review]

**Comment by NinjaPhenix** [comment]
Have to agree with haykam on discord that the add interface, whilst a good addition, should probably not be included in the access widener library. A dedicated byte code manipulation library would probably be a good idea, could eventually expose additional features too e.g. adding fields ( this should be one of the easier things to do ), methods, ect.

**Comment by shartte** [comment]
I did write up a small gist that should demonstrate the kind of simplification (both for users, and for API maintainers) that global AWs and interface injection brings to the table:

https://gist.github.com/shartte/760c7fce30d81cd248c2f91005565613

File: src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java
Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java]: Done

File: src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java
Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java]: Done

File: src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java
Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java]: Done

File: src/main/java/net/fabricmc/accesswidener/RemappingDecorator.java
Comment (modmuss50) [inline src/main/java/net/fabricmc/accesswidener/RemappingDecorator.java:25]: AccessWidenerRemapper would be a better classname imo

**Review by modmuss50 — CHANGES_REQUESTED** [review]
Looks really good. In a few places error handing/logic should be a bit more offensive imo. If something isnt right it should bail out.

Please let me know if you have any questions or dont like what ive suggested.

Many thanks for this :)

File: src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java
Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java:264]: How so?

Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java:264]: (I'll add a test)

File: src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java
Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java:265]: Hey, it was you who wanted strict mode 😁 I would have enabled this all the time. (This was probably for loader btw.)
Should I do that?

File: src/main/java/net/fabricmc/accesswidener/AccessWidenerWriter.java
Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerWriter.java:86]: This isn't guessing though. It uses the minimal version number needed to represent the features that are used.
In my opinion, passing the format version through to the visitor breaks encapsulation of the format details.

File: src/main/java/net/fabricmc/accesswidener/ForwardingVisitor.java
Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/ForwardingVisitor.java:26]: One could pass an array here and modify it on the outside, the clone is defensive.

File: src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java
Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java:269]: Agreed

File: src/main/java/net/fabricmc/accesswidener/RemappingDecorator.java
Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/RemappingDecorator.java:50]: The remapper provider was intended to support the case where you reuse the same visitor chain for multiple readers.
I do that in loom by reading multiple mod's AWs into the same visitor chain to produce an collated AW for the transformation.
It turned out that all mod AWs are in intermediary anway, so it wasn't really an issue, but I honestly didn't know that when I wrote this 😁

File: src/main/java/net/fabricmc/accesswidener/RemappingDecorator.java
Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/RemappingDecorator.java:25]: I'll rename it. Although I somewhat liked the theme of "chaining" decorators and filters together.

File: src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java
Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java:269]: What do we do with the existing "AccessWidenerVisitor" though....

Comment (Juuxel) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java:269]: Isn’t it the ASM class visitor? I feel that can be renamed to AccessWidenerClassVisitor.

Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java:269]: That'd work for me, should I just go ahead with that?

File: src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java
Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java:264]: Discussed this with player on discord, I don't think that class names in class-files can contain periods, so this should be safe, I think.

File: src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java
Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java:265]: Removed strict mode, enabled class-name check by default.

File: src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java
Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java:262]: I'd do both actually.

But, "cannot find class a.b.C" will not be a good error message here, if the class should be a/b/C.

File: src/main/java/net/fabricmc/accesswidener/AccessWidenerWriter.java
Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerWriter.java:65]: Done

File: src/main/java/net/fabricmc/accesswidener/AccessWidenerWriter.java
Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerWriter.java:86]: Read through it again. I'd keep this, but make it clearer. It allows merging different AWs that use different format versions, which is something I find appealing.

File: src/main/java/net/fabricmc/accesswidener/AccessWidenerWriter.java
Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerWriter.java:110]: Done

File: src/main/java/net/fabricmc/accesswidener/AccessWidenerWriter.java
Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerWriter.java:135]: Done

File: src/main/java/net/fabricmc/accesswidener/AccessWidenerWriter.java
Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerWriter.java:160]: Done

File: src/main/java/net/fabricmc/accesswidener/AccessWidenerWriter.java
Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerWriter.java:98]: Renamed to ClassRule, introduced base class Rule. (Other suggestion would be "Line").

File: src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java
Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java:269]: Done.

File: src/main/java/net/fabricmc/accesswidener/RemappingDecorator.java
Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/RemappingDecorator.java:25]: Done

File: src/main/java/net/fabricmc/accesswidener/RemappingDecorator.java
Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/RemappingDecorator.java:50]: I introduced a secondary constructor in case only a 1:1 remapping is needed, but kept the other one.

File: src/main/java/net/fabricmc/accesswidener/AccessWidenerWriter.java
Comment (modmuss50) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerWriter.java:86]: What about setting the destired version in the consturctor, you should be able to merge v1's into v2, but not the other way around. The writer should be set to a specific version, and fail if it needs to write something spcific to v2 when in v1 writing mode.

Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerWriter.java:86]: The code that initiates the merging wouldn't really know, though.
I mean the other option is to just always write the newest version (v2 in this case).

File: src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java
Comment (liach) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java:264]: I think this is totally safe, too. I recall JVM refuses to load a class that refers to a class name entry in constant pool with `.` in it.

File: src/main/java/net/fabricmc/accesswidener/AccessWidenerWriter.java
Comment (liach) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerWriter.java:86]: We should always write the newest version.

Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerWriter.java:86]: I've changed it to use the highest version that was piped into the writer.
When remapping a single AW and writing it out, it'll stay at that version.
When remapping and merging multiple AWs, it'll use the highest version of the input AWs.

Comment (modmuss50) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerWriter.java:86]: I dont see these changes?

Comment (modmuss50) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerWriter.java:86]: Oh, found them they are on another branch on your fork.

Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerWriter.java]: Sorry about that, pushed the updates to the correct branch now. This also included fixes to other review comments.

**Comment by modmuss50** [comment]
For the writer I was more thinking something along the line of:

```java
public final class AccessWidenerWriter implements AccessWidenerVisitor {
	private final Writer writer;
	private final int version;

	private String namespace;

	public AccessWidenerWriter(Writer writer, int version) {
		this.writer = writer;
		this.version = version;
	}

	public AccessWidenerWriter(Writer writer) {
		this(writer, 2 /* Latest version */);
	}
	
	@Override
	public void visitHeader(String namespace) throws IOException {
		if (this.namespace == null) {
			writer.write(String.format("accessWidener\tv%d\t%s\n", this.version, namespace));
		} else if (!this.namespace.equals(namespace)) {
			throw new IllegalArgumentException("Cannot write different namespaces to the same file ("
					+ this.namespace + " != " + namespace + ")");
		}
		
		this.namespace = namespace;
	}

	@Override
	public void visitClass(String name, AccessWidenerReader.AccessType access, boolean transitive) throws IOException {
		writeAccess(access, transitive);
		writer.write(String.format("\tclass\t%s\n", name));
	}

	@Override
	public void visitMethod(String owner, String name, String descriptor, AccessWidenerReader.AccessType access, boolean transitive) throws IOException {
		writeAccess(access, transitive);
		writer.write(String.format("\tmethod\t%s\t%s\t%s\n", owner, name, descriptor));
	}

	@Override
	public void visitField(String owner, String name, String descriptor, AccessWidenerReader.AccessType access, boolean transitive) throws IOException {
		writeAccess(access, transitive);
		writer.write(String.format("\tfield\t%s\t%s\t%s\n", owner, name, descriptor));
	}
	
	private void writeAccess(AccessWidenerReader.AccessType access, boolean transitive) throws IOException {
		if (transitive) {
			if (version < 2) {
				throw new IllegalStateException("Cannot write transitive rule in version " + version);
			}

			writer.write("transitive-");
		}

		writer.write(access.toString());
	}
}
```

**Comment by shartte** [comment]
@modmuss50 That'll break merging from multiple readers into the same writer since it'd write multiple headers with potentially different versions, breaking the format.

Using this approach, we'd also need to add a way to the reader to read only the version number of an access widener so the writer can be constructed. This complicates the code in loom (and loader), and I still don't understand the benefit of doing it.

I could see the issue with the previous version of the writer "magically" downgrading the version, but that's now no longer the case.

*anyway:* If this is blocking merging this, I'll get on it and change it.

**Comment by modmuss50** [comment]
> since it'd write multiple headers with potentially different versions, breaking the format.

No, it wont. It will only write one header with the version that it was setup to write.

> we'd also need to add a way to the reader to read only the version number

Yes thats correct, it should be very easy and fast to do, as only the first few bytes needs to be read. I fail to see how the code is going to become complex (comparied to what the code is already doing), loader can always use the latest version.

Before you go and change everything I think it would be good to get @sfPlayer1 to take a look and provide some input on this.

**Comment by shartte** [comment]
Oh you're right, I missed the if that guards writing the header.

It's not the reader code that's complex (other than the tiny duplication).

I had the loom side envisioned to work like this for remapping the Mod's AW to intermediary.
```
public byte[] getRemappedAccessWidener(Remapper asmRemapper, String targetNamespace) throws IOException {
	AccessWidenerWriter writer = new AccessWidenerWriter();
	RemappingDecorator remapper = new RemappingDecorator(writer, asmRemapper, "named", targetNamespace);
	new AccessWidenerReader(remapper).read(modAccessWidener);
	return writer.write();
}
```

The changed variant is minimally longer, which isn't really a concern, but I fail to understand the benefit of this approach:

```
public byte[] getRemappedAccessWidener(Remapper asmRemapper, String targetNamespace) throws IOException {
        int version = AccessWidenerReader.read(modAccessWidener);
        ByteArrayOutputStream bos = new ByteArrayOutputStream();
        try (OutputStreamWriter osWriter = new OutputStreamWriter(bos, AccessWidenerReader.ENCODING)) {
        	AccessWidenerWriter writer = new AccessWidenerWriter(osWriter, version);
        	RemappingDecorator remapper = new RemappingDecorator(writer, asmRemapper, "named", targetNamespace);
        	new AccessWidenerReader(remapper).read(modAccessWidener);
        }

	return bos.toByteArray();
}
```

**Comment by modmuss50** [comment]
@shartte I spoke to player and we agreed with an approach similar to what I posted above. It saves having to collect up the rules to write them later. The version is only relavent to the reader and writer so shouldnt be passed via the visitor (I know I suggested this earlier but I take that back). Passing the version in feels a bit more concreate, I feel like I can be more sure with whats going to come out.

File: src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java
Comment (modmuss50) [inline src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java:60]: the method should throw the IOException

File: src/main/java/net/fabricmc/accesswidener/AccessWidenerRemapper.java
Comment (modmuss50) [inline src/main/java/net/fabricmc/accesswidener/AccessWidenerRemapper.java:30]: Why would this ever be null, if it doesnt need remapping dont use the AccessWidenerRemapper

File: src/main/java/net/fabricmc/accesswidener/AccessWidenerRemapper.java
Comment (modmuss50) [inline src/main/java/net/fabricmc/accesswidener/AccessWidenerRemapper.java:65]: I dont see the need of the provider? Just seems to make it neednlessly complex for something we arent going to use?

**Review by modmuss50 — CHANGES_REQUESTED** [review]
Looking good, just some minor things outside of the writer.

File: src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java
Comment (shartte) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java:60]: But why? The caller isn't passing a stream or reader that would throw. Any IOException being thrown here is a bug in the reader, not the caller's fault. (StringReader never throws, BufferedReader should only throw on programming bugs).

Comment (modmuss50) [reply src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java:60]: Ok, sure 👍 its not going to happen anyway.

**Comment by shartte** [comment]
I am going to make those changes but still think it makes the library worse.

> The version is only relavent to the reader and writer so shouldnt be passed via the visitor

We are now leaking the version to clients of the library (loom + loader).

**Review by sfPlayer1 — APPROVED** [review]
LGTM, thanks!

**Review by modmuss50 — APPROVED** [review]
👍

**Pull request merged**
